### PR TITLE
Fix ROOT libraries dependencies 

### PR DESCRIPTION
**Pull request by linev**
Extraction from #5181 

First of all - add libraries in dependency list when class from such library used
Second - try to remove includes which are not used and coming from libraries which are not in dep list
In two places try to use virtual  interfaces (TVirtualTreePlayer and TVirtualPacketizer) to avoid linking to implementation library.

### CODE REVIEW

File: roofit/roofitcore/CMakeLists.txt [hunk 1]
```diff
@@ -460,7 +460,9 @@ ROOT_STANDARD_LIBRARY_PACKAGE(RooFitCore
     RIO
     MathCore
     Foam
-  LINKDEF inc/LinkDef.h
+    Smatrix
```
Comment (hageboeck) [inline roofit/roofitcore/CMakeLists.txt]: I grepped for Smatrix, and didn't find anything in RooFit. What makes you think that `roofitcore` uses it?

Comment (linev) [reply roofit/roofitcore/CMakeLists.txt]: @hageboeck Seems to be, you are right. I just checked with code from #5181 - it works also without `Smatrix`. Probably I misinterpret some missing include which were coming from `MathCore`, not from Smatrix. I will extract rootfit changes in separate PR

Comment (linev) [reply roofit/roofitcore/CMakeLists.txt]: I extract rootfit changes in #5232

**Comment by vgvassilev** [comment]
Forgot to say, this should be only 2 commits: one the dependencies the other the header/linkdef adjustments to be able to semantically track such improvements...

**Comment by linev** [comment]
@pcanal Can you approve my changes in `TreePlayer`. I extend virtual interface to be able use `TVirtualTreePlayer` class from GUI

File: graf2d/win32gdk/src/TGWin32.cxx [hunk 2]
```diff
@@ -4802,6 +4801,11 @@ Window_t TGWin32::CreateWindow(Window_t parent, Int_t x, Int_t y,
    GdkColor background_color;
    ULong_t xmask = 0;
 
+   /* copy these bits from TGFrame.h, otherwise one should link against gui lib */
+   UInt_t kMainFrame       = BIT(0);
```
Comment (vgvassilev) [inline graf2d/win32gdk/src/TGWin32.cxx]: Don’t forget to move those in a common place.

Comment (linev) [reply graf2d/win32gdk/src/TGWin32.cxx]: Is it necessary? Probably we should ask @bellenot - that is better solution here.

Comment (bellenot) [reply graf2d/win32gdk/src/TGWin32.cxx]: What about TVirtualX? (as a quick idea...)

Comment (linev) [reply graf2d/win32gdk/src/TGWin32.cxx]: @bellenot It is your decision. I could copy related  enum from TGFrame.h to core/gui/inc/GuiTypes.h - there are many other similar enums.

Comment (bellenot) [reply graf2d/win32gdk/src/TGWin32.cxx]: Fine with me

Comment (bellenot) [reply graf2d/win32gdk/src/TGWin32.cxx]: @linev BTW, just wondering what does that implies for users' code? Will they have to change their includes?

Comment (linev) [reply graf2d/win32gdk/src/TGWin32.cxx]: @bellenot It is not a user problem - it is missing libs dependency in ROOT itself. Via this enum `win32gdk` have to have `Gui` in list of dependencies while it uses includes from `Gui`. It comes out when I tried to compile ROOT using only source dirs in include paths

Comment (bellenot) [reply graf2d/win32gdk/src/TGWin32.cxx]: @linev I know that, and I agree, but I was just wondering if moving those declarations around could break users code at some point...

Comment (linev) [reply graf2d/win32gdk/src/TGWin32.cxx]: Only if somebody, who do not include `TGFrame.h` before, uses `EFrameType` as enum or any constants from it. Therefore my effort with `less_includes` PR should pay here - normally `GuiTypes.h` should not appear in non-GUI classes.

Comment (linev) [reply graf2d/win32gdk/src/TGWin32.cxx]: Here is my commit: https://github.com/root-project/root/pull/5231/commits/a9688b568eb358184ddbdaacf7b9a852a14d19d4

Comment (bellenot) [reply graf2d/win32gdk/src/TGWin32.cxx]: Very good! Thanks @linev !

**Pull request merged**